### DISCUSSION TITLE
아이템 37. ordinal 인덱싱 대신 EnumMap을 사용하라

### DISCUSSION THREAD

**Discussion by chikeem90** [Study-2-Effective-Java/Effective-Java#96] [3. 과제]
### EnumMap이란?
- 열거 타입을 키로 사용하도록 설계한 Map 구현체

### 예제1. 식물리스트를 생애주기별 집합으로 묶을 때
1. 배열과 ordinal을  사용하는 경우
```java
public class Main {
    public static void main(String[] args) {
        Plant[] garden = {
            new Plant("바질", LifeCycle.ANNUAL),
            new Plant("캐러웨이", LifeCycle.BIENNIAL),
            new Plant("딜", LifeCycle.ANNUAL),
            new Plant("라벤더", LifeCycle.PERENNIAL),
            new Plant("파슬리", LifeCycle.BIENNIAL),
            new Plant("로즈마리", LifeCycle.PERENNIAL)
        };
        // length = 3
        Set<Plant>[] plantsByLifeCycle = (Set<Plant>[]) new Set[LifeCycle.values().length];
        // 각 index별 hashSet 할당
        for (int i = 0; i < plantsByLifeCycle.length; i++) {
            plantsByLifeCycle[i] = new HashSet<>();
        }
        // ordinal에 해당하는 index의 hashSet에 add
        for (Plant p : garden) {
            plantsByLifeCycle[p.lifeCycle.ordinal()].add(p);
        }
        // 각 배열 index별로 무슨 값을 가지는지 알 수 없기 때문에,, labeling도 해야 함
        for (int i = 0; i < plantsByLifeCycle.length; i++) {
            System.out.printf("%s: %s%n", LifeCycle.values()[i], plantsByLifeCycle[i]);
        }
    }
}
```
- ordinal을 사용해서 각 배열의 원소에 접근하기 때문에 각 배열의 인덱스가 어떤 열거 타입을 대표하는지 명확하지 않고, 만약에 열거 타입이 추가되거나 삭제되는 경우 데이터가 꼬이거나 ArrayIndexOutOfBoundsException이 발생할 가능성이 있음.

2. EnumMap을 사용하는 경우
- LifeCycle을 key로 가지고 Set<Plant>를 value로 갖는 EnumMap을 적용하면 더 명료하고 안전하고 성능도 배열을 사용할 때와 비등함.
```java
public class Main {
    public static void main(String[] args) {
        Plant[] garden = {
            new Plant("바질", LifeCycle.ANNUAL),
            new Plant("캐러웨이", LifeCycle.BIENNIAL),
            new Plant("딜", LifeCycle.ANNUAL),
            new Plant("라벤더", LifeCycle.PERENNIAL),
            new Plant("파슬리", LifeCycle.BIENNIAL),
            new Plant("로즈마리", LifeCycle.PERENNIAL)
        };

        Map<LifeCycle, Set<Plant>> plantsByLifeCycle = new EnumMap<>(LifeCycle.class);
        for (LifeCycle lc : LifeCycle.values()) {
            plantsByLifeCycle.put(lc, new HashSet<>());
        }
        for (Plant p : garden) {
            plantsByLifeCycle.get(p.lifeCycle).add(p);
        }
        System.out.println(plantsByLifeCycle);
    }
}
```

### 예제2. 상태 전이 매핑 프로그램
1. 배열과 ordinal을 사용하는 경우
```java
public enum Phase {
    SOLID, LIQUID, GAS;

    public enum Transition {
        MELT, FREEZE, BOIL, CONDENSE, SUBLIME, DEPOSIT;

        private static final Transition[][] TRANSITIONS = {
            { null, MELT, SUBLIME },
            { FREEZE, null, BOIL },
            { DEPOSIT, CONDENSE, null }
        };

        public static Transition from(Phase from, Phase to) {
            return TRANSITIONS[from.ordinal()][to.ordinal()];
        }
    }
}
```
- 상태 전이에 매핑되는 값을 메뉴얼하게 관리해야함.
- Phase나 Transition에 상태 값이 추가되거나 삭제되는 경우 TRANSITIONS도 직접 변경해야함.

2. EnumMap을 사용하는 경우
```java
public enum Phase {
    SOLID, LIQUID, GAS;

    public enum Transition {
        MELT(SOLID, LIQUID), FREEZE(LIQUID, SOLID), BOIL(LIQUID, GAS),
        CONDENSE(GAS, LIQUID), SUBLIME(SOLID, GAS), DEPOSIT(GAS, SOLID);

        private final Phase from;
        private final Phase to;

        Transition(Phase from, Phase to) {
            this.from = from;
            this.to = to;
        }

        private static final Map<Phase, Map<Phase, Transition>> m = Stream.of(Transition.values())
            .collect(Collectors.groupingBy(t -> t.from, () -> new EnumMap<>(Phase.class),
                Collectors.toMap(t -> t.to, t -> t, (x, y) -> y,() -> new EnumMap<>(Phase.class))));

        public static Transition from(Phase from, Phase to) {
            return m.get(from).get(to);
        }
    }
}
```
- Phase나 Transition 유형이 추가되더라도 추가로 변경해야할 부분 X.

### HashMap을 써도 되는거 아닌가?
- HashMap의 경우 put, get 할 때 key에 대해 매번 hash function을 수행하지만, EnumMap의 경우 내부적으로 단순히 열거 타입별 배열을 사용하여 구현하고 있음.
- 그렇기 때문에 key가 열거 타입일 경우 EnumMap을 사용하는 것이 더 효율적,,

```java
    public EnumMap(Class<K> keyType) {
        this.keyType = keyType;
        keyUniverse = getKeyUniverse(keyType);
        vals = new Object[keyUniverse.length];
    }
```

**Comment by JoisFe** [comment]
EnumMap이 좋은건 알겠는데 HashMap이 있는데 왜 굳이 EnumMap을 쓰지?? 라는 의문이 있었는데 정리해주셔서 감사합니다.  좀 더 이해를 하기 위해 알아보니

```EnumMap의 경우 내부적으로 단순히 열거 타입별 배열을 사용하여 구현``` --> 해시 맵과 달리 조회하는데 O(1)이 되지 않아 조회가 느리지 않을까? 라는 의문을 가지고 EnumMap과 HashMap 차이를 좀 더 알아봤습니다.

## EnumMap vs HashMap
### EnumMap이 HashMap보다 좋은점
1. 성능이 좋음
조회 시간 복잡도가 O(1)인 해시맵보다 EnumMap이 어떻게 더 좋을지 이해가 가지 않았습니다만 알아보니 Hash 경우 검색이 매우 빠르지만 이를 위해 해싱 작업이 필요한데 enum 경우 단일 객체임을 보장!! 할 수 있기에 이러한 작업이 필요하지 않다고 합니다.
(단 항상 EnumMap이 HashMap보다 빠르다고는 보장할 수 없음)

### EnumMap
시간 복잡도
get, containsKey, next : O(1) 

내부적으로 단순히 열거 타입별 배열이 어떻게 해시 알고리즘을 쓰지 않고 HashMap과 같이 EnumMap 또한 조회가 O(1)이 가능한지 궁금했습니다.

<img width="692" alt="image" src="https://user-images.githubusercontent.com/90208100/215729585-5706c502-1196-4b83-bd9d-0ce382ff7701.png">

EnumMap get 메서드의 구현을 보면 단순히 해당 키의 ordinal 메서드의 반환값을 인덱스로 사용해 배열에서 값을 꺼내 리턴하는 것을 볼 수 있습니다.

값에 접근할때 인덱스로 직접 배열에 접근하기에 CRUD 모두 O(1)이 되는 것 이었습니다.

- EnumMap, HashMap 둘다 get은 O(1)
- Enum은 get 연산을 항상 O(1) 보장
- HashMap은 해시 충돌이 발생하면 O(1)을 보장할 수 없음 (충돌이 잦으면 O(logN)에 근접)


2. 순서를 기억
가끔 HashMap은 put을 하고 나서 순서를 보장하지 않아 LinkedHashMap을 사용하는 경우가 존재한데 EnumMap 경우 순서를 기억한다고 합니다.


정리 감사합니다.!!
